source/common/router/header_parser.cc

```diff
@@ -351,19 +358,34 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
       switch (entry.append_action_) {
         PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
       case HeaderValueOption::APPEND_IF_EXISTS_OR_ADD:
-        headers.addReferenceKey(key, value);
+        headers_to_add.emplace_back(key, value);
         break;
       case HeaderValueOption::ADD_IF_ABSENT:
         if (auto header_entry = headers.get(key); header_entry.empty()) {
-          headers.addReferenceKey(key, value);
+          headers_to_add.emplace_back(key, value);
```

Comment:
I think this changes the semantics, and I think there's a need for more tests and docs clarification of the desired behavior.

Previous semantics:
APPEND_IF_EXISTS_OR_ADD["a", "val1"]
ADD_IF_ABSENT["a", "val2"]
will result with {"a"->"val1"}

New semantics:
APPEND_IF_EXISTS_OR_ADD["a", "val1"]
ADD_IF_ABSENT["a", "val2"]
will result with {"a"->"val1,val2"}

Similarly multiple ADD_IF_ABSENT["a", "val1"], ADD_IF_ABSENT["a", "val2"]

Reply:
Taking a step back, I wonder if the right way to look at this is to reject a config that has multiple headers_to_add_ keys with the same value (i.e., it should be a map in the first place).

Reply:
I think that the same problem can be recreated with multiple headers too, not with only one:
start with: `a = "aaaaaa", b="bbbbbb"`

set b to `%REQ(a)%%REQ(a)%` -> b is aaaaaaaaaaaa
add new header c as `%REQ(b)%%REQ(b)%`. It should be bbbbbbbbbbbb but it will be aaaaaaaaaaaaaaaaaaaaaaaa

Reply:
> I think this changes the semantics, and I think there's a need for more tests and docs clarification of the desired behavior.
> 
> Previous semantics:
> APPEND_IF_EXISTS_OR_ADD["a", "val1"]
> ADD_IF_ABSENT["a", "val2"]
> will result with {"a"->"val1"}
> 
> New semantics:
> APPEND_IF_EXISTS_OR_ADD["a", "val1"]
> ADD_IF_ABSENT["a", "val2"]
> will result with {"a"->"val1,val2"}
> 
> Similarly multiple ADD_IF_ABSENT["a", "val1"], ADD_IF_ABSENT["a", "val2"]

To be honest, I am not sure if we should worry about that, because it is kind of "artificial" config, which specifies multiple values for the same header, which in practice should never happen. IMHO the only thing we should be concerned here is that a wrong config does not lead to CPU/memory overload or exhaustion (what is done in this PR). In the worst case a user may disable runtime guard to go back to the original behaviour. I think that your suggestion to catch things at the config time (allow only one modifier per header name) is much better idea. WDYT @adisuissa ?

Reply:
I think rejecting is a better approach if it doesn't break compatibility with the previous version of this code.

Reply:
Modified config processing code to reject multiple OVERWRITE_IF_EXISTS_OR_ADD for the same header. This simplifies code when doing actual header evaluation.